presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -0,0 +1,301 @@
+
+        queryManagementExecutor = Executors.newScheduledThreadPool(queryManagerConfig.getQueryManagerExecutorPoolSize(), threadsNamed("query-management-%s"));
+        queryManagementExecutorMBean = new ThreadPoolExecutorMBean((ThreadPoolExecutor) queryManagementExecutor);
+    }
+
+    public void start()
+    {
+        queryManagementExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                failAbandonedQueries();
+            }
+            catch (Throwable e) {
+                log.error(e, "Error cancelling abandoned queries");
+            }
+
+            try {
+                enforceTimeLimits();
+            }
+            catch (Throwable e) {
+                log.error(e, "Error enforcing query timeout limits");
+            }
+
+            try {
+                removeExpiredQueries();
+            }
+            catch (Throwable e) {
+                log.error(e, "Error removing expired queries");
+            }
+
+            try {
+                pruneExpiredQueries();
+            }
+            catch (Throwable e) {
+                log.error(e, "Error pruning expired queries");
+            }
+        }, 1, 1, TimeUnit.SECONDS);
+    }
+
+    public void stop()
+    {
+        boolean queryCancelled = false;
+        for (T trackedQuery : queries.values()) {
+            if (trackedQuery.isDone()) {
+                continue;
+            }
+
+            log.info("Server shutting down. Query %s has been cancelled", trackedQuery.getQueryId());
+            trackedQuery.fail(new PrestoException(SERVER_SHUTTING_DOWN, "Server is shutting down. Query " + trackedQuery.getQueryId() + " has been cancelled"));
+            queryCancelled = true;
+        }
+        if (queryCancelled) {
+            try {
+                TimeUnit.SECONDS.sleep(5);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+        queryManagementExecutor.shutdownNow();
+    }
+
+    public Collection<T> getAllQueries()
+    {
+        return ImmutableList.copyOf(queries.values());
+    }
+
+    public T getQuery(QueryId queryId)
+            throws NoSuchElementException
+    {
+        return tryGetQuery(queryId)
+                .orElseThrow(NoSuchElementException::new);
+    }
+
+    public Optional<T> tryGetQuery(QueryId queryId)
+    {
+        requireNonNull(queryId, "queryId is null");
+        return Optional.ofNullable(queries.get(queryId));
+    }
+
+    public boolean addQuery(T execution)
+    {
+        return queries.putIfAbsent(execution.getQueryId(), execution) == null;
+    }
+
+    /**
+     * Query is finished and expiration should begin.
+     */
+    public void expireQuery(QueryId queryId)
+    {
+        tryGetQuery(queryId)
+                .ifPresent(expirationQueue::add);
+    }
+
+    @Managed(description = "Query garbage collector executor")
+    @Nested
+    public ThreadPoolExecutorMBean getManagementExecutor()
```

Comment:
How would this appear in debugging/ops tools? What would the object name be?

Reply:
Actually, this is totally wrong.  QueryTracker is not exported to JMX, so it won't show up at all.  I'm going to change query tracker to take the executor in the constructor, and will use the existing executor in the `SqlQueryManager`, so there will be no change to the existing JMX exports.